shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -62,15 +62,52 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
       env: MutMap[Str, TypeInfo],
       mthEnv: MutMap[(Str, Str) \/ (Opt[Str], Str), MethodType],
       lvl: Int,
+      qenv: MutMap[Str, SkolemTag], // * SkolemTag for variables in quasiquotes
+      fvars: MutSet[ST], // * Free variables
+      inQuote: Bool, // * Is in quasiquote
       inPattern: Bool,
       tyDefs: Map[Str, TypeDef],
       tyDefs2: MutMap[Str, DelayedTypeInfo],
       inRecursiveDef: Opt[Var], // TODO rm
       extrCtx: ExtrCtx,
   ) {
-    def +=(b: Str -> TypeInfo): Unit = env += b
+    def +=(b: Str -> TypeInfo): Unit = {
+      env += b
+      // * For lambdas and let bindings, we should not have bindings with the same name in one context.
```

Comment:
That assumption is incorrect: `let x = 1; let x = 2; x`. Unless I misunderstood what's written?

Reply:
Yes. It is misleading. It should denote "quoted lambdas and let bindings". I've updated the comments.

Reply:
BTW I still don't understand what the comment is trying to say.

Reply:
```js
x `=> x `=> ......
//    [ctx1 [ctx2 ]]
```
In the above example, even though we have two parameters called `x`, their skolem tags are stored in different contexts. Assume that the first `x` has skolem tag `a1` and the second `x` has `a2`. `a1` is stored in `ctx1.quoteSkolemEnv` (the body of the first lambda) while `a2` is stored in `ctx2.quoteSkolemEnv` (the body of the second lambda, and the first `x` will be shadowed). So each `quoteSkolemEnv` should not have two bindings with the same name.
The only possible cases where a `quoteSkolemEnv` has two bindings with the same name are caused by temp variables desugared from UCS:
```js
// shared/src/test/diff/qq/Codegen.mls
:js
:ne
code"if 3 is
  1 then 114
  2 then 514
  _ then 0
"
//│ Code[0 | 114 | 514, ??tmp0 & ~??tmp00]
//│ // Prelude
//│ class TypingUnit24 {}
//│ const typing_unit24 = new TypingUnit24;
//│ // Query 1
//│ res = ((tmp01) => Let(Var(tmp01), IntLit(3), ((tmp01) => Let(Var(tmp01), IntLit(3), CaseOf(Var(tmp01), Case(IntLit(1), IntLit(114), Case(IntLit(2), IntLit(514), Wildcard(IntLit(0)))))))(freshName("tmp0"))))(freshName("tmp0"));
//│ // End of generated code
```
In the above example, we would have two different skolems without the check, and therefore there is an extrusion, which is not correct. We should use the same skolem for those desugared bindings and they are always bound.

Reply:
How does this have anything to do with the UCS? In this case the desugared term is:
```scala
let tmp0 = 3 in let tmp0 = 3 in case tmp0 of { 1 => 114; 2 => 514; _ => 0 }
```
You can write that code by hand as well, and in both cases, the inner context will be different from the outer one and there shouldn't be any confusion.

Reply:
It happens in `typeArms`:
```scala
val (req_ty, bod_ty, (tys, rest_ty)) = scrutVar match {
  case S(v) =>
    if (newDefs) {
      newCtx += v.name -> VarSymbol(tagTy & patTy, v)
```
We can do the same process for the `newCtx` as we do for `Lam` and `Let` but the `con` function is only available in `typeTerm`.

Reply:
> It happens in `typeArms`:

What's "it"? 

And this code has nothing to do with the UCS. This is the part of the code that refines scrutinee variables. It is not related to the shadowing of variables that happen in the example you gave above. 

> We can do the same process for the `newCtx`

What process are you talking about? Can you try to be clearer and more specific when you write comments like that?

Reply:
I mean that this refinement invokes the `+=` of `newCtx`, which will create a new skolem for `tmp0` if we do not have the check `!quoteSkolemEnv.contains(b._1)` because the whole expression is inside a quote. That's why we got the unexpected extrusion above.

One potential solution is like this:
```scala
val newCtx = if (ctx.inQuote) ctx.enterQuotedScope else ctx.nest
val (req_ty, bod_ty, (tys, rest_ty)) = scrutVar match {
  case S(v) =>
    if (newDefs) {
      newCtx += v.name -> VarSymbol(tagTy & patTy, v)
      val bod_ty = typeTerm(bod)(newCtx, raise, vars, genLambdas)
      if (ctx.inQuote) {
        val ctxTy = freshVar(noTyProv, N)(ctx.lvl)
        ctx.trackFVs(con(newCtx.getCtxTy, newCtx.getAllQuoteSkolemsWith(ctxTy), ctxTy)(ctx))
      }
// ...
```
The problem is that we can not access `con` in `typeArms`.

Reply:
> The problem is that we can not access `con` in `typeArms`.

Function `con` is just a very small helper method which doesn't really matter. You can just call `constrain` directly!!